src/intervals/construction.jl

```diff
@@ -64,71 +64,68 @@ struct Interval{T} <: Real
-Interval(a) = Interval(a, a)
-Interval(a::Tuple) = Interval(a...)
-
-Interval(a::T, b::S) where {T<:AbstractFloat, S} = Interval{promote_type(T, S)}(a, b)
-Interval(a::T, b::S) where {T, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
-Interval(a::T, b::S) where {T<:AbstractFloat, S<:AbstractFloat} = Interval{promote_type(T, S)}(a, b)
-Interval(a::T, b::T) where {T<:AbstractFloat} = Interval{T}(a, b)
-
-Interval(a::T, b::S) where {T, S} = Interval{promote_type(default_bound(), T, S)}(a, b)
-Interval(a::T, b::T) where T = Interval{promote_type(default_bound(), T)}(a, b)
-
-#= Irrational =#
-# Single argument Irrational constructor are in IntervalArithmetic.jl
-# as generated functions need to be define last.
-Interval(a::Irrational, b::T) where {T<:AbstractFloat} = Interval{T}(T(a, RoundDown), b)
-Interval(a::T, b::Irrational) where {T<:AbstractFloat} = Interval{T}(a, T(b, RoundUp))
-
-function Interval(a::Irrational, b::Irrational)
-    T = default_bound()
-    return Interval{T}(a, b)
-end
-
-#= Interval =#
-Interval(x::Interval) = x
-Interval{T}(x::Interval{T}) where T = x
-Interval{T}(x::Interval) where T = Interval{T}(inf(x), sup(x))
-
-#= Complex =#
-Interval(x::Complex) = Interval(real(x)) + im*Interval(imag(x))
+Interval{T}(a, b) where {T} = Interval{T}(T(a, RoundDown), T(b, RoundUp))
 
-# These definitions have been put there because generated functions must be
-# defined after all methods they use.
-Interval(x::Irrational) = Interval{default_bound()}(x)
-
-@generated function Interval{T}(x::Irrational) where T
-    res = Interval{T}(x(), x())  # Precompute the interval
-    return :(return $res)  # Set body of the function to return the precomputed result
-end
+Interval{T}(x::Interval) where {T} = Interval{T}(inf(x), sup(x))
 
 # promotion
 
-Base.promote_rule(::Type{Interval{T}}, ::Type{Interval{S}}) where {T,S} =
+Base.promote_rule(::Type{Interval{T}}, ::Type{Interval{S}}) where {T, S} =
     Interval{promote_type(T, S)}
 
 """
-    interval(a, b)
+    interval(::Type{T}, a, b)
 
 Create an interval, checking whether [a, b] is a valid `Interval`
-If so, then an `Interval(a, b)` object is returned;
+If so, then an `Interval{T}(a, b)` object is returned;
 if not, a warning is printed and the empty interval is returned.
 """
-function interval(a::T, b::S) where {T<:Real, S<:Real}
-    is_valid_interval(a, b) && return Interval(a, b)
+function interval(::Type{T}, a, b) where {T}
+    is_valid_interval(a, b) && return Interval{T}(a, b)
     @warn "Invalid input, empty interval is returned"
-    return emptyinterval(promote_type(T, S))
+    return emptyinterval(T)
+end
+interval(a::T, b::S) where {T<:AbstractFloat, S<:AbstractFloat} =
+    interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T<:AbstractFloat, S} = interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T, S<:AbstractFloat} = interval(promote_type(T, S), a, b)
+interval(a::T, b::S) where {T, S} = interval(promote_type(default_bound(), T, S), a, b)
```

Comment:
I kept this because it was the behaviour before the PR.

When we restrict the parametric type to `Union{Rational,AbstractFloat}`, I think we should actually prevent this from happening; what do you think?